### PR TITLE
extmod/ssl: Add SSLContext 

### PR DESCRIPTION
**Pull request by Carglglz**
This is a proof of concept/demo of what was proposed in #8915 and #8252 to get closer to CPython `ssl.py` module:
This implements a very basic version of `SSLContext` in MicroPython and adds new features in C:

- `ssl.SSLContext()`:
    - set_ciphers()
    - get_ciphers()
    - load_verify_locations()
    - load_cert_chain()
    - wrap_socket()
    - check_hostname
    - verify_mode
    - protocol
    - load_default_certs() 
    - reset()
- `_SSLSocket.cipher()`
- Certificate verification errors as `ValueError:` + string indicating failure reasons.

I also made a demo test `tests/net_inet/test_ssl_context_client.py` that works in CPython, unix and esp32 port.
There is still some questions that need to be resolved (see comments in `extmod/ssl/ssl.py`), but I think this is good enough to get the ball rolling 👀

### CODE REVIEW

**Comment by Carglglz** [comment]
Well I'm not sure if this is the right way to implement this, but it is a start. (C code probably needs some cleaning/fixing 🤷🏼‍♂️)

I added a `multi_net/ssl_context_rsa.py` test and 
I tried to add `MBEDTLS_VERSION` but it seems it is only available in `MBEDTLS 3.x`, that is why I introduced `# MICROPY_SSL_MBEDTLS_3` macro, and why some unix port builds fails.

In my unix port build this works,
```python
MicroPython v1.19.1-219-g8e33c1bd1-dirty on 2022-07-28; darwin [GCC 4.2.1] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import ssl
>>> ssl.MBEDTLS_VERSION
'mbed TLS 3.2.1'
```
But `esp-idf 4.x`  uses `MBEDTLS 2.x`.

Other than that, tests passes in `esp32` too 👍🏼 .

**Comment by Carglglz** [comment]
I've just added `SSLContext.reset()` method and a `http_server_ssl_context.py` example. 
This allows to "reuse" the SSLContext. I guess this is not CPython compliant but I'm not sure how to balance between memory allocation and context reusability.
Right now what context does (or I think it does) is:
1) Context initiation --> memory allocation --> load cert/key/cadata/conf
2) Context wrap socket --> creates a ssl socket that use the allocated memory from context 
3) SSLSocket is closed and frees the allocated memory

So to reuse the context, it needs to be reinitiated to allocate memory again, or another option would be to stop the SSLSocket from freeing the allocated memory (but I'm not sure this is the way go... 🤔 )

**Comment by Carglglz** [comment]
Note about `ussl.wrap_socket` deprecation:

I "deprecated" `ussl.wrap_socket` (just commented out for now) in favour of `ssl.wrap_socket` that is backwards compatible with `ussl.wrap_socket` and with the docs api description (so it accepts kargs `key`, `cert`, or `keyfile`,  `certfile`)

**Comment by Carglglz** [comment]
Quick recap of things that need to be solved yet:
   - esp32 port:
       - Preallocate buffers for SSLsockets #8940 
       - Fix for esp-idf `mbedtls/config.h` to enable certs datetime validation #9089
 
Other than that I've just added uasyncio+tls following #5840 in the latest commit since CPython uses `SSLContext` here as well see [asyncio.loop.create_connection](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_connection)  
I've also test this with `microdot_asyncio`  and it works too, so let me know if you want to implement any of this 👍🏼.

**Comment by Carglglz** [comment]
@dpgeorge I think this is ready for review, I could squash some commits or split this PR if it is too big, whatever makes it easier.

The key part is of course `extmod/mod_usslmbedtls.c` and `extmod/ssl`, the rest of changes/additions are to integrate `ssl` module and `SSLContext` in `unix`, `esp32` and`rp2` ports and `uasyncio` with the corresponding tests and examples (docs are on the way...maybe in a different PR?) 

Despite all the efforts there are still some key differences with `CPython` that I couldn't figure out, i.e. how to implement them in a reasonable way without being too complex or cumbersome... these are:
- `CPython` can wrap sockets without being connected, and does it in a way that I'm not sure how it can be replicated in `MicroPython`
> For client-side sockets, the context construction is lazy; if the underlying socket isn’t connected yet, the context construction will be performed after connect() is called on the socket. For server-side sockets, if the socket has no remote peer, it is assumed to be a listening socket, and the server-side SSL wrapping is automatically performed on client connections accepted via the accept() method.

see [ssl.py](https://github.com/python/cpython/blob/main/Lib/ssl.py#L990)


- `mbedtls` vs `OpenSSL` cipher naming conventions as I mentioned [here](https://github.com/micropython/micropython/pull/9506#issuecomment-1269180374)
- create default context / load default certs since there is no default certs embedded in firmware (or how to load them without enabling `PEM`, as we already [discussed](https://github.com/micropython/micropython/pull/8252#discussion_r924162202) )

Also on a second thought #8940 would be nice to have but it's not necessary for this to work in `esp32` port...

**tldr** I think this is ready to go 👍🏼

File: extmod/modssl_mbedtls.c [hunk 1]
```diff
@@ -63,6 +74,21 @@ typedef struct _mp_obj_ssl_socket_t {
...
+typedef struct _mp_obj_ssl_context_t {
+    mp_obj_base_t base;
+    mp_obj_t sock;
+    mbedtls_entropy_context entropy;
+    mbedtls_ctr_drbg_context ctr_drbg;
+    mbedtls_ssl_context ssl;
+    mbedtls_ssl_config conf;
+    mbedtls_x509_crt cacert;
+    mbedtls_x509_crt cert;
+    mbedtls_pk_context pkey;
```
Comment (dpgeorge) [inline extmod/modssl_mbedtls.c:87]: All this context/state has been added here (in the SSLContext object) but it also remains in the SSLSocket object.  I don't think that's the correct approach because (1) it uses additional memory, to duplicate the state in each SSLSocket, (2) it means that a multiple SSLSockets can't share state from the same SSLContext.

I think the only thing that should be in the SSLSocket struct is `mbedtls_ssl_context` (the naming is poor, but `mbedtls_ssl_config` maps to SSLContext and `mbedtls_ssl_context` maps to SSLSocket).

File: extmod/ssl/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+from .ssl import *
```
Comment (dpgeorge) [inline extmod/ssl/__init__.py:1]: This file is not doing anything, it's not in the manifest (and isn't really needed).

File: extmod/ssl/ssl.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    return ctx.wrap_socket(
+        sock,
+        server_side=server_side,
+        cert_reqs=cert_reqs,
+        server_hostname=server_hostname,
+        do_handshake=do_handshake,
+    )
+
+
+class SSLContext:
```
Comment (dpgeorge) [inline extmod/ssl/ssl.py:43]: I think this should inherit from the native "ssl context" class, and add methods that are missing, rather than have a `ctx` member.

File: extmod/ssl/ssl.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        sock,
+        server_side=False,
+        do_handshake_on_connect=True,
+        server_hostname=None,
+    ):
+        if self.check_hostname and server_hostname is None:
+            raise ValueError("check_hostname requires server_hostname")
+        # to be substituted by e.g. _ussl._context_wrap_socket in modussl_mbedtls.c ?:
+        # _ussl._context_wrap_socket(*args, **kargs)
+        if self._reload_ctx:
```
Comment (dpgeorge) [inline extmod/ssl/ssl.py:158]: I think this `_reload_ctx` logic can be removed if the mbedtls state is properly split between SSLContext and SSLSocket (as discussed above).

**Comment by dpgeorge** [comment]
Thanks @Carglglz for working on this and keeping it up to date (rebased).

In addition to the above comments, another issue with this PR is that it's doing too much.  From what I can tell it is:
- reworking the existing ssl module to introduce `ssl.SSLContext`
- adding a lot of new methods and functionality to `SSLContext`
- adding time support to certificates
- adding support for SSL connections in `asyncio`
- adding tests and examples

Don't get me wrong, these are all good features to have.  But it's hard to review when there is so much going on.  And in this case a lot of the components are independent enough that they could be in separate PRs.

**Comment by dpgeorge** [comment]
See #11862 for an alternative implementation of just `ssl.SSLContext` (that's not intending to replace this whole PR).

**Comment by Carglglz** [comment]
> Don't get me wrong, these are all good features to have. But it's hard to review when there is so much going on. And in this case a lot of the components are independent enough that they could be in separate PRs.

Yes I know 😅 , that's exactly what I proposed here 👍🏼 
> @dpgeorge I think this is ready for review, I could squash some commits or split this PR if it is too big, whatever makes it easier.

see full comment in https://github.com/micropython/micropython/pull/8968#issuecomment-1397704920 

I will review the changes you proposed, but in general I'm OK as long as functionality is kept. Probably the best approach is to rebase this PR on #11862 and split it into smaller PRs  based on this:
>  PR0: https://github.com/micropython/micropython/pull/11862
PR1: adding a lot of new methods and functionality to SSLContext
PR2: adding time support to certificates
PR3: adding support for SSL connections in asyncio
PR4: adding tests and examples

**Comment by dpgeorge** [comment]
> I will review the changes you proposed, but in general I'm OK as long as functionality is kept.

OK, thanks for being flexible with how we implement this.  Since this SSL stuff is fundamental to a lot of other components that will be built on top of it, we need to make sure it's done well from the beginning.

If you could review #11862 that would be great.

**Comment by Carglglz** [comment]
Closing this in favour of #11862 #11888 #11896 #11897
